dialer.go

```diff
@@ -52,26 +57,67 @@ func NewDialer(ctx context.Context, opts ...DialerOption) (*Dialer, error) {
 	for _, opt := range opts {
 		opt(&cfg)
 	}
-
 	client, err := sqladmin.NewService(context.Background(), cfg.sqladminOpts...)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create sqladmin client: %v", err)
 	}
+
+	dialCfg := dialCfg{
+		ipType:       cloudsql.IP_TYPE_PUBLIC,
+		tcpKeepAlive: defaultTCPKeepAlive,
+	}
+	for _, opt := range cfg.dialOpts {
+		opt(&dialCfg)
+	}
+
 	d := &Dialer{
-		instances: make(map[string]*cloudsql.Instance),
-		sqladmin:  client,
-		key:       key,
+		instances:      make(map[string]*cloudsql.Instance),
+		sqladmin:       client,
+		key:            key,
+		defaultDialCfg: dialCfg,
 	}
 	return d, nil
 }
 
 // Dial creates an authorized connection to a Cloud SQL instance specified by it's instance connection name.
-func (d *Dialer) Dial(ctx context.Context, instance string) (net.Conn, error) {
+func (d *Dialer) Dial(ctx context.Context, instance string, opts ...DialOption) (net.Conn, error) {
+	cfg := d.defaultDialCfg
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
 	i, err := d.instance(instance)
 	if err != nil {
 		return nil, err
 	}
-	return i.Connect(ctx)
+	ipAddrs, tlsCfg, err := i.ConnectInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	addr, ok := ipAddrs[cfg.ipType]
+	if !ok {
+		return nil, fmt.Errorf("instance '%s' does not have IP of type '%s'", instance, cfg.ipType)
+	}
+	addr = net.JoinHostPort(addr, "3307")
+
+	conn, err := proxy.Dial(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	if c, ok := conn.(*net.TCPConn); ok {
+		if err := c.SetKeepAlive(true); err != nil {
+			return nil, fmt.Errorf("failed to set keep-alive: %w", err)
```

Comment:
Interesting.  Would you suggest just returning the text then?

This code is actually the same as it was before, I just yanked it from the `cloudsql.Instance.Connect` to here, since it seemed better to keep the logic related to connecting in the "dialer" file.

Reply:
I think returning the text makes the most sense in which case the format specifier becomes `%v`.